File: backend/services/gemini_ipo_service.py
"""
Gemini AI IPO Service for Backend
Fetches and processes IPO data using Google's Gemini AI from real sources like Groww and Zerodha
"""

import asyncio
import aiohttp
import json
import logging
from datetime import datetime, timedelta
from typing import List, Dict, Optional
import os
import re
from bs4 import BeautifulSoup
import google.generativeai as genai
from utils.logger import get_logger
from utils.cache import cache_manager

logger = get_logger(__name__)

class GeminiIPOService:
    def __init__(self):
        self.api_key = os.getenv('GEMINI_API_KEY')
        self.model = None
        self.cache_timeout = 30 * 60  # 30 minutes
        self.is_initialized = False
        self.last_fetch = None
        self.daily_task = None
        self.last_daily_update = None
        self.session = None

    async def initialize(self):
        """Initialize the Gemini service"""
        if self.is_initialized:
            return True

        if not self.api_key:
            logger.warning("Gemini API key not found. Please set GEMINI_API_KEY environment variable")
            return False

        try:
            # Configure the API key
            genai.configure(api_key=self.api_key)
            
            # Create HTTP session for web scraping
            self.session = aiohttp.ClientSession(
                timeout=aiohttp.ClientTimeout(total=30),
                headers={
                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
                }
            )
            
            self.is_initialized = True
            logger.info("✅ Gemini IPO Service initialized successfully")
            return True

        except Exception as e:
            logger.error(f"❌ Failed to initialize Gemini IPO Service: {str(e)}")
            return False

    async def fetch_ipo_data(self):
        """Fetch IPO data and enhance with Gemini AI"""
        if not self.is_initialized:
            # Try to initialize if not already done
            await self.initialize()

        try:
            # Use the consolidated real-time service to get base data
            from services.real_time_ipo_service import real_time_ipo_service
            all_ipos = await real_time_ipo_service.fetch_all_ipo_data()
            
            if not all_ipos:
                logger.warning("No real IPO data found from sources")
                return []
            
            # Enhance with Gemini AI
            enhanced_ipos = await self.enhance_with_gemini(all_ipos)
            
            # Cache the results
            cache_manager.cache.set('gemini_ipo_data', enhanced_ipos, self.cache_timeout)
            self.last_fetch = datetime.utcnow()
            
            logger.info(f"✅ Fetched and enhanced {len(enhanced_ipos)} IPOs")
            return enhanced_ipos

        except Exception as e:
            logger.error(f"❌ Error fetching IPO data: {str(e)}")
            return []

    async def fetch_current_ipos(self):
        """Alias for fetch_ipo_data for backward compatibility"""
        return await self.fetch_ipo_data()

    def get_service_status(self):
        """Get the current service status"""
        return {
            "is_initialized": self.is_initialized,
            "has_api_key": bool(self.api_key),
            "last_fetch": self.last_fetch.isoformat() if self.last_fetch else None,
            "last_daily_update": self.last_daily_update.isoformat() if self.last_daily_update else None,
            "daily_updates_running": bool(self.daily_task and not self.daily_task.done()),
            "service": "Gemini AI with Real-Time IPO Service"
        }

    def get_status(self):
        """Alias for get_service_status for backward compatibility"""
        return self.get_service_status()

    async def get_market_sentiment(self):
        """Get market sentiment analysis using real data"""
        if not self.is_initialized:
            raise Exception("Service not initialized")

        try:
            # Get current IPO data for sentiment analysis
            current_ipos = await self.fetch_current_ipos()
            
            # Calculate sentiment based on real data
            if not current_ipos:
                return {
                    "sentiment_score": 5.0,
                    "analysis": "No current IPO data available for sentiment analysis.",
                    "key_drivers": ["Market data unavailable"]
                }
            
            # Analyze IPO performance metrics
            total_ipos = len(current_ipos)
            positive_gmp_count = sum(1 for ipo in current_ipos if ipo.get('current_gmp', 0) > 0)
            avg_gmp = sum(ipo.get('current_gmp', 0) for ipo in current_ipos) / total_ipos
            
            # Count oversubscribed IPOs
            oversubscribed_count = sum(1 for ipo in current_ipos 
                                     if 'oversubscribed' in ipo.get('subscription_status', '').lower())
            
            # Calculate sentiment score (1-10)
            base_score = 5.0
            
            # Adjust based on GMP performance
            if avg_gmp > 20:
                base_score += 2.0
            elif avg_gmp > 10:
                base_score += 1.0
            elif avg_gmp < 0:
                base_score -= 1.5
            
            # Adjust based on positive GMP ratio
            positive_ratio = positive_gmp_count / total_ipos
            if positive_ratio > 0.8:
                base_score += 1.0
            elif positive_ratio < 0.4:
                base_score -= 1.0
            
            # Adjust based on subscription levels
            if oversubscribed_count > total_ipos * 0.6:
                base_score += 0.5
            
            # Ensure score is within bounds
            sentiment_score = max(1.0, min(10.0, base_score))
            
            # Generate analysis text
            if sentiment_score >= 7.5:
                sentiment_text = "bullish"
                market_condition = "strong investor confidence"
            elif sentiment_score >= 6.0:
                sentiment_text = "cautiously optimistic"
                market_condition = "selective investor interest"
            elif sentiment_score >= 4.0:
                sentiment_text = "neutral"
                market_condition = "mixed market signals"
            else:
                sentiment_text = "bearish"
                market_condition = "investor caution prevailing"
            
            analysis = f"Market sentiment is {sentiment_text} with {market_condition}. Average GMP of ₹{avg_gmp:.1f} across {total_ipos} IPOs, with {positive_gmp_count} showing positive premiums."
            
            # Key drivers based on data
            key_drivers = []
            if avg_gmp > 15:
                key_drivers.append("Strong grey market premiums")
            if oversubscribed_count > 0:
                key_drivers.append("High subscription levels")
            if positive_ratio > 0.7:
                key_drivers.append("Broad-based investor interest")
            else:
                key_drivers.append("Selective stock picking")
            
            if not key_drivers:
                key_drivers = ["Market consolidation", "Investor caution", "Sector rotation"]
            
            return {
                "sentiment_score": round(sentiment_score, 1),
                "analysis": analysis,
                "key_drivers": key_drivers
            }

        except Exception as e:
            logger.error(f"❌ Error getting market sentiment: {str(e)}")
            return {
                "sentiment_score": 6.0,
                "analysis": "Market sentiment analysis temporarily unavailable. System showing moderate optimism based on recent IPO trends.",
                "key_drivers": ["System analysis", "Recent IPO performance", "Market trends"]
            }

    async def start_daily_updates(self):
        """Start daily automatic updates"""
        if self.daily_task and not self.daily_task.done():
            logger.info("Daily updates already running")
            return

        async def daily_update_task():
            while True:
                try:
                    await asyncio.sleep(24 * 60 * 60)  # Wait 24 hours
                    await self.fetch_ipo_data()
                    self.last_daily_update = datetime.utcnow()
                    logger.info("✅ Daily IPO data update completed")
                except Exception as e:
                    logger.error(f"❌ Daily update failed: {str(e)}")

        self.daily_task = asyncio.create_task(daily_update_task())
        logger.info("✅ Daily updates started")

    def stop_daily_updates(self):
        """Stop daily automatic updates"""
        if self.daily_task and not self.daily_task.done():
            self.daily_task.cancel()
            logger.info("✅ Daily updates stopped")

    async def close(self):
        """Close the service and cleanup resources"""
        if self.session:
            await self.session.close()
        self.stop_daily_updates()

# Create global service instance
gemini_ipo_service = GeminiIPOService()

File: src/lib/geminiIPOService.js
// Gemini AI-powered IPO Data Service
// Fetches real-time IPO data using Google's Gemini AI

class GeminiIPOService {
  constructor() {
    this.apiKey = process.env.NEXT_PUBLIC_GEMINI_API_KEY || process.env.GEMINI_API_KEY
    this.baseUrl = 'https://generativelanguage.googleapis.com/v1beta/models/gemini-pro:generateContent'
    this.cache = new Map()
    this.cacheTimeout = 30 * 60 * 1000 // 30 minutes cache
    this.lastFetch = null
    this.isInitialized = false
    this.dailyUpdateInterval = null
    this.lastDailyUpdate = null
    this.autoUpdateEnabled = true
  }

  // Initialize the service
  async initialize() {
    if (this.isInitialized) return true

    if (!this.apiKey) {
      console.warn('Gemini API key not found. Please set NEXT_PUBLIC_GEMINI_API_KEY or GEMINI_API_KEY')
      return false
    }

    try {
      // Test the API connection
      await this.testConnection()
      this.isInitialized = true
      console.log('✅ Gemini IPO Service initialized successfully')
      return true
    } catch (error) {
      console.error('❌ Failed to initialize Gemini IPO Service:', error.message)
      return false
    }
  }

  // Test API connection
  async testConnection() {
    const response = await fetch(`${this.baseUrl}?key=${this.apiKey}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        contents: [{
          parts: [{
            text: 'Hello, are you working?'
          }]
        }]
      })
    })

    if (!response.ok) {
      throw new Error(`API test failed: ${response.status} ${response.statusText}`)
    }

    return true
  }

  // Fetch current IPO data using Gemini AI with web search
  async fetchCurrentIPOs() {
    try {
      if (!await this.initialize()) {
        throw new Error('Gemini service not initialized')
      }

      // Check cache first
      const cacheKey = 'current_ipos'
      if (this.isCacheValid(cacheKey)) {
        console.log('📋 Returning cached IPO data')
        return this.cache.get(cacheKey).data
      }

      console.log('🤖 Fetching current IPO data from Gemini AI with web search...')

      const prompt = `
        Search the internet and provide REAL, current IPO information for the Indian stock market as of ${new Date().toDateString()}.

        Please search for and find:
        1. IPOs currently OPEN for subscription (check BSE, NSE, SEBI websites)
        2. IPOs UPCOMING in the next 30 days (announced by companies)
        3. Recently CLOSED IPOs (last 7 days)
        4. Current Grey Market Premium (GMP) data from reliable sources

        Search these sources:
        - BSE India (bseindia.com)
        - NSE India (nseindia.com) 
        - SEBI official announcements
        - Chittorgarh.com for GMP data
        - IPOWatch.in
        - Economic Times IPO section
        - Moneycontrol IPO section
        - Business Standard IPO news

        For each REAL IPO found, provide:
        - Company name (exact name)
        - Issue price range (in rupees)
        - Issue size (in crores)
        - Subscription dates (open and close dates)
        - Expected listing date
        - Current Grey Market Premium (GMP) if available
        - Industry/sector
        - Board type (Main Board or SME)
        - Brief company description
        - Current status (Open/Upcoming/Closed)

        IMPORTANT INSTRUCTIONS:
        - ONLY include REAL IPOs that you can verify from official sources
        - Search the internet for current, up-to-date information
        - Verify GMP data from grey market sources
        - If no real IPOs are currently active, return an empty array []
        - Do NOT create fake or example data
        - Cross-reference information from multiple sources for accuracy

        Format the response as a JSON array:
        [
          {
            "name": "Actual Company Name IPO",
            "company": "Actual Company Name",
            "priceRange": "₹100 - ₹120",
            "issueSize": 1500,
            "openDate": "2024-01-15",
            "closeDate": "2024-01-17", 
            "listingDate": "2024-01-25",
            "gmp": 25,
            "gmpPercent": 15.5,
            "status": "Open/Upcoming/Closed",
            "industry": "Technology",
            "boardType": "Main Board",
            "description": "Brief verified company description",
            "lotSize": 100,
            "verified": true,
            "sources": ["BSE", "NSE", "Chittorgarh"]
          }
        ]

        Please search the web thoroughly and provide only verified, real IPO data.
      `

      const response = await this.callGeminiAPI(prompt)
      const ipos = this.parseIPOResponse(response)

      // Cache the results
      this.cache.set(cacheKey, {
        data: ipos,
        timestamp: Date.now()
      })

      this.lastFetch = new Date().toISOString()
      console.log(`✅ Fetched ${ipos.length} real IPOs from Gemini AI web search`)

      return ipos

    } catch (error) {
      console.error('❌ Error fetching IPOs from Gemini:', error.message)
      return this.getFallbackData()
    }
  }

  // Get detailed IPO analysis using Gemini AI
  async getIPOAnalysis(ipoName) {
    try {
      if (!await this.initialize()) {
        throw new Error('Gemini service not initialized')
      }

      const prompt = `
        Provide a detailed analysis of the "${ipoName}" IPO including:

        1. Company Background:
           - Business model and operations
           - Key products/services
           - Market position

        2. Financial Analysis:
           - Revenue trends
           - Profitability
           - Debt levels
           - Use of IPO proceeds

        3. Market Analysis:
           - Industry outlook
           - Competition
           - Growth prospects

        4. Risk Assessment:
           - Key risks and challenges
           - Market risks
           - Company-specific risks

        5. Investment Recommendation:
           - Should investors apply?
           - Expected listing gains
           - Long-term prospects
           - Risk level (Low/Medium/High)

        Please provide factual, current information and format as JSON:
        {
          "company": "Company Name",
          "analysis": {
            "background": "...",
            "financials": "...",
            "market": "...",
            "risks": "...",
            "recommendation": "Buy/Hold/Avoid",
            "expectedGains": "10-15%",
            "riskLevel": "Medium",
            "confidence": 0.8
          }
        }
      `

      const response = await this.callGeminiAPI(prompt)
      return this.parseAnalysisResponse(response)

    } catch (error) {
      console.error('❌ Error getting IPO analysis:', error.message)
      return null
    }
  }

  // Get market sentiment and trends
  async getMarketSentiment() {
    try {
      if (!await this.initialize()) {
        throw new Error('Gemini service not initialized')
      }

      const prompt = `
        Analyze the current IPO market sentiment in India as of ${new Date().toDateString()}.

        Please provide:
        1. Overall market sentiment (Bullish/Bearish/Neutral)
        2. Recent IPO performance trends
        3. Investor appetite for new issues
        4. Market factors affecting IPO pricing
        5. Upcoming IPO pipeline outlook

        Format as JSON:
        {
          "sentiment": "Bullish/Bearish/Neutral",
          "trends": "...",
          "investorAppetite": "High/Medium/Low",
          "marketFactors": ["factor1", "factor2"],
          "outlook": "...",
          "confidence": 0.8,
          "lastUpdated": "${new Date().toISOString()}"
        }
      `

      const response = await this.callGeminiAPI(prompt)
      return this.parseSentimentResponse(response)

    } catch (error) {
      console.error('❌ Error getting market sentiment:', error.message)
      return null
    }
  }

  // Get GMP updates for specific IPOs
  async getGMPUpdates(ipoNames = []) {
    try {
      if (!await this.initialize()) {
        throw new Error('Gemini service not initialized')
      }

      const ipoList = ipoNames.length > 0 ? ipoNames.join(', ') : 'all current IPOs'

      const prompt = `
        Get the latest Grey Market Premium (GMP) data for ${ipoList} in the Indian market.

        For each IPO, provide:
        - Current GMP in rupees
        - GMP percentage
        - Trend (increasing/decreasing/stable)
        - Last updated time

        Format as JSON array:
        [
          {
            "name": "Company Name IPO",
            "gmp": 25,
            "gmpPercent": 15.5,
            "trend": "increasing",
            "lastUpdated": "2024-01-15T10:30:00Z",
            "source": "grey market"
          }
        ]

        Only include real, verified GMP data. If no GMP data is available, return empty array.
      `

      const response = await this.callGeminiAPI(prompt)
      return this.parseGMPResponse(response)

    } catch (error) {
      console.error('❌ Error getting GMP updates:', error.message)
      return []
    }
  }

  // Call Gemini API
  async callGeminiAPI(prompt) {
    const response = await fetch(`${this.baseUrl}?key=${this.apiKey}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        contents: [{
          parts: [{
            text: prompt
          }]
        }],
        generationConfig: {
          temperature: 0.1, // Low temperature for factual responses
          topK: 1,
          topP: 0.8,
          maxOutputTokens: 2048,
        }
      })
    })

    if (!response.ok) {
      throw new Error(`Gemini API error: ${response.status} ${response.statusText}`)
    }

    const data = await response.json()
    
    if (!data.candidates || !data.candidates[0] || !data.candidates[0].content) {
      throw new Error('Invalid response from Gemini API')
    }

    return data.candidates[0].content.parts[0].text
  }

  // Parse IPO response from Gemini
  parseIPOResponse(response) {
    try {
      // Extract JSON from response
      const jsonMatch = response.match(/\[[\s\S]*\]/)
      if (!jsonMatch) {
        console.log('No JSON array found in response, checking for empty array indication')
        if (response.toLowerCase().includes('no real ipos') || 
            response.toLowerCase().includes('empty array') ||
            response.includes('[]')) {
          return []
        }
        throw new Error('No valid JSON found in response')
      }

      const ipos = JSON.parse(jsonMatch[0])
      
      // Validate and enhance IPO data
      return ipos.map((ipo, index) => ({
        id: `gemini_${Date.now()}_${index}`,
        name: ipo.name || `${ipo.company} IPO`,
        company: ipo.company || ipo.name?.replace(' IPO', ''),
        priceRange: ipo.priceRange || 'TBA',
        issueSize: ipo.issueSize || 0,
        gmp: ipo.gmp || 0,
        gmpPercent: ipo.gmpPercent || 0,
        status: ipo.status || 'Unknown',
        isProfitable: (ipo.gmp || 0) >= 20 || (ipo.gmpPercent || 0) >= 10,
        openDate: ipo.openDate || this.getRandomFutureDate(1, 30),
        closeDate: ipo.closeDate || this.getRandomFutureDate(3, 35),
        listingDate: ipo.listingDate || this.getRandomFutureDate(10, 45),
        industry: ipo.industry || 'Others',
        boardType: ipo.boardType || 'Main Board',
        description: ipo.description || `${ipo.company} is a company in the ${ipo.industry} sector.`,
        lotSize: ipo.lotSize || this.calculateLotSize(ipo.priceRange),
        confidenceScore: 0.9, // High confidence for Gemini data
        source: 'gemini_ai',
        lastUpdated: new Date().toISOString(),
        recommendation: this.generateRecommendation(ipo.gmp || 0, ipo.gmpPercent || 0),
        riskLevel: this.calculateRiskLevel(ipo.gmp || 0, ipo.gmpPercent || 0)
      }))

    } catch (error) {
      console.error('Error parsing IPO response:', error.message)
      return []
    }
  }

  // Parse analysis response
  parseAnalysisResponse(response) {
    try {
      const jsonMatch = response.match(/\{[\s\S]*\}/)
      if (!jsonMatch) {
        throw new Error('No valid JSON found in analysis response')
      }

      return JSON.parse(jsonMatch[0])
    } catch (error) {
      console.error('Error parsing analysis response:', error.message)
      return null
    }
  }

  // Parse sentiment response
  parseSentimentResponse(response) {
    try {
      const jsonMatch = response.match(/\{[\s\S]*\}/)
      if (!jsonMatch) {
        throw new Error('No valid JSON found in sentiment response')
      }

      return JSON.parse(jsonMatch[0])
    } catch (error) {
      console.error('Error parsing sentiment response:', error.message)
      return null
    }
  }

  // Parse GMP response
  parseGMPResponse(response) {
    try {
      const jsonMatch = response.match(/\[[\s\S]*\]/)
      if (!jsonMatch) {
        return []
      }

      return JSON.parse(jsonMatch[0])
    } catch (error) {
      console.error('Error parsing GMP response:', error.message)
      return []
    }
  }

  // Utility functions
  isCacheValid(key) {
    const cached = this.cache.get(key)
    if (!cached) return false
    return Date.now() - cached.timestamp < this.cacheTimeout
  }

  getRandomFutureDate(minDays, maxDays) {
    const today = new Date()
    const randomDays = Math.floor(Math.random() * (maxDays - minDays + 1)) + minDays
    const futureDate = new Date(today)
    futureDate.setDate(today.getDate() + randomDays)
    return futureDate.toISOString().split('T')[0]
  }

  calculateLotSize(priceRange) {
    if (!priceRange || priceRange === 'TBA') return 100
    
    const match = priceRange.match(/₹(\d+)/)
    const price = match ? parseInt(match[1]) : 100
    
    if (price <= 200) return 75
    if (price <= 500) return 30
    if (price <= 1000) return 15
    return 10
  }

  generateRecommendation(gmp, gmpPercent) {
    if (gmp >= 50 || gmpPercent >= 25) return 'Strong Buy'
    if (gmp >= 20 || gmpPercent >= 10) return 'Buy'
    if (gmp >= 0 || gmpPercent >= 0) return 'Hold'
    return 'Avoid'
  }

  calculateRiskLevel(gmp, gmpPercent) {
    if (gmp < 0 || gmpPercent < 0) return 'High'
    if (gmp < 20 || gmpPercent < 10) return 'Medium'
    return 'Low'
  }

  getFallbackData() {
    // Return empty array when Gemini fails
    console.log('📋 Returning empty data due to Gemini API failure')
    return []
  }

  // Clear cache
  clearCache() {
    this.cache.clear()
    console.log('🗑️ Gemini IPO Service cache cleared')
  }

  // Start daily automatic updates when service starts
  startDailyUpdates() {
    if (this.dailyUpdateInterval) {
      console.log('Daily updates already running')
      return
    }

    console.log('🔄 Starting daily automatic IPO updates from Gemini AI...')

    // Initial fetch
    this.fetchCurrentIPOs()

    // Calculate time until next 9 AM IST
    const scheduleNextUpdate = () => {
      const now = new Date()
      const nextUpdate = new Date()
      nextUpdate.setHours(9, 0, 0, 0) // 9 AM

      // If it's already past 9 AM today, schedule for tomorrow
      if (now.getHours() >= 9) {
        nextUpdate.setDate(nextUpdate.getDate() + 1)
      }

      const timeUntilUpdate = nextUpdate.getTime() - now.getTime()

      setTimeout(() => {
        console.log('🌅 Daily update triggered at 9 AM')
        this.performDailyUpdate()

        // Schedule next update (every 24 hours)
        this.dailyUpdateInterval = setInterval(() => {
          this.performDailyUpdate()
        }, 24 * 60 * 60 * 1000)
      }, timeUntilUpdate)
    }

    scheduleNextUpdate()
    console.log('✅ Daily updates scheduled for 9 AM IST')
    
    // Auto-start daily updates when initialized
    if (this.autoUpdateEnabled) {
      this.autoUpdateEnabled = true
    }
  }

  // Stop daily automatic updates
  stopDailyUpdates() {
    if (this.dailyUpdateInterval) {
      clearInterval(this.dailyUpdateInterval)
      this.dailyUpdateInterval = null
      console.log('⏹️ Stopped daily automatic updates')
    }
  }

  // Perform daily update
  async performDailyUpdate() {
    try {
      console.log('🔄 Performing daily IPO data update from Gemini AI...')
      
      // Clear cache to force fresh fetch
      this.clearCache()
      
      // Fetch fresh data from the internet
      const ipos = await this.fetchCurrentIPOs()
      
      this.lastDailyUpdate = new Date().toISOString()
      
      console.log(`✅ Daily update complete: ${ipos.length} IPOs fetched from web`)
      
      // Emit update event if there are listeners
      if (typeof window !== 'undefined') {
        window.dispatchEvent(new CustomEvent('gemini-daily-update', {
          detail: { ipos, timestamp: this.lastDailyUpdate }
        }))
      }
      
      return ipos
    } catch (error) {
      console.error('❌ Daily update failed:', error.message)
      return []
    }
  }

  // Force immediate update
  async forceUpdate() {
    console.log('🔄 Forcing immediate IPO data update...')
    this.clearCache()
    return await this.fetchCurrentIPOs()
  }

  // Get service status
  getStatus() {
    return {
      isInitialized: this.isInitialized,
      hasApiKey: !!this.apiKey,
      lastFetch: this.lastFetch,
      lastDailyUpdate: this.lastDailyUpdate,
      cacheSize: this.cache.size,
      autoUpdateEnabled: this.autoUpdateEnabled,
      dailyUpdatesRunning: !!this.dailyUpdateInterval,
      service: 'Gemini AI with Web Search'
    }
  }
}

// Export singleton instance
const geminiIPOService = new GeminiIPOService()
export default geminiIPOService